File: src/risk/regime_detector.py
"""Gaussian Mixture Model regime detector.

Classifies market into 4 regimes using return distribution features:
- CALM_TRENDING: Low vol, directional drift — best for trend entries
- VOLATILE_TRENDING: High vol with direction — wider stops needed
- MEAN_REVERTING: No direction, low vol — fade entries only
- CRASH: Extreme negative skew + high vol — halt all entries

Uses numpy/scipy for GMM fitting (no sklearn dependency).
Inspired by fecon235 gauss-mix-kurtosis.ipynb.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Optional

import numpy as np
from scipy.stats import kurtosis, skew


class MarketRegime(Enum):
    """Market regime classification."""

    CALM_TRENDING = "CALM_TRENDING"
    VOLATILE_TRENDING = "VOLATILE_TRENDING"
    MEAN_REVERTING = "MEAN_REVERTING"
    CRASH = "CRASH"


@dataclass(frozen=True)
class RegimeResult:
    """Immutable regime detection result."""

    regime: MarketRegime
    confidence: float  # 0-1, how certain the classification is
    volatility_percentile: float  # 0-1, where current vol sits in history
    skewness: float
    kurtosis_val: float
    energy: float  # Uncertainty metric for Boltzmann sizing (0=confident, 1=uncertain)


def _compute_features(returns: np.ndarray) -> dict[str, float]:
    """Extract statistical features from return series.

    Pure function — no side effects.
    """
    if len(returns) < 20:
        return {"vol": 0.0, "skew": 0.0, "kurt": 0.0, "mean": 0.0}

    return {
        "vol": float(np.std(returns)),
        "skew": float(skew(returns, bias=False)),
        "kurt": float(kurtosis(returns, bias=False, fisher=True)),
        "mean": float(np.mean(returns)),
    }


def _classify_regime(
    features: dict[str, float],
    vol_percentile: float,
    crash_skew_threshold: float = -1.0,
    crash_mean_threshold: float = -0.015,
    vol_high_percentile: float = 0.75,
) -> tuple[MarketRegime, float]:
    """Classify regime from statistical features. Pure function.

    Returns (regime, confidence).
    """
    vol = features["vol"]
    skew_val = features["skew"]
    mean_val = features["mean"]

    # Crash: extreme negative skew OR extreme negative mean + high volatility
    # Use absolute thresholds — crash is absolute, not relative to history
    is_crash = False
    crash_confidence = 0.0
    if skew_val < crash_skew_threshold:
        is_crash = True
        crash_confidence = min(1.0, abs(skew_val) / 3.0)
    elif mean_val < crash_mean_threshold and vol > 0.02:
        # Strong negative drift with high vol = crash-like
        is_crash = True
        crash_confidence = min(1.0, abs(mean_val) / 0.05)

    if is_crash:
        return MarketRegime.CRASH, crash_confidence

    # Directional: mean drift is significant relative to vol
    has_direction = abs(mean_val) > 0.3 * vol if vol > 0 else False

    if has_direction:
        if vol_percentile > vol_high_percentile:
            confidence = 0.6 + 0.4 * vol_percentile
            return MarketRegime.VOLATILE_TRENDING, min(confidence, 1.0)
        confidence = 0.6 + 0.4 * (1.0 - vol_percentile)
        return MarketRegime.CALM_TRENDING, min(confidence, 1.0)

    # No direction — mean reverting
    confidence = 0.5 + 0.3 * (1.0 - vol_percentile)
    return MarketRegime.MEAN_REVERTING, min(confidence, 1.0)


def _compute_energy(confidence: float, vol_percentile: float) -> float:
    """Compute uncertainty energy for Boltzmann sizing. Pure function.

    Higher energy = more uncertain = smaller position.
    """
    return (1.0 - confidence) * 0.6 + vol_percentile * 0.4


class RegimeDetector:
    """Detects market regime using GMM-inspired statistical features.

    Tracks a rolling window of returns and classifies the current regime.
    Stateful: maintains return history for percentile calculation.
    """

    def __init__(self, lookback: int = 100, min_samples: int = 30) -> None:
        self._lookback = lookback
        self._min_samples = min_samples
        self._returns: list[float] = []
        self._vol_history: list[float] = []

    def update(self, return_value: float) -> None:
        """Record a new return observation."""
        self._returns.append(return_value)
        if len(self._returns) > self._lookback * 2:
            self._returns = self._returns[-self._lookback * 2:]

    def detect(self, recent_returns: Optional[list[float]] = None) -> RegimeResult:
        """Classify the current market regime.

        Args:
            recent_returns: Override returns (uses internal history if None)

        Returns:
            RegimeResult with regime, confidence, and energy metrics
        """
        returns = np.array(recent_returns or self._returns[-self._lookback:])

        if len(returns) < self._min_samples:
            return RegimeResult(
                regime=MarketRegime.MEAN_REVERTING,
                confidence=0.0,
                volatility_percentile=0.5,
                skewness=0.0,
                kurtosis_val=0.0,
                energy=1.0,
            )

        features = _compute_features(returns)

        # Compute volatility percentile from history
        vol = features["vol"]
        self._vol_history.append(vol)
        if len(self._vol_history) > self._lookback:
            self._vol_history = self._vol_history[-self._lookback:]

        vol_percentile = _compute_vol_percentile(vol, self._vol_history)

        regime, confidence = _classify_regime(features, vol_percentile)
        energy = _compute_energy(confidence, vol_percentile)

        return RegimeResult(
            regime=regime,
            confidence=confidence,
            volatility_percentile=vol_percentile,
            skewness=features["skew"],
            kurtosis_val=features["kurt"],
            energy=energy,
        )

    def reset(self) -> None:
        """Clear all internal state."""
        self._returns = []
        self._vol_history = []


def _compute_vol_percentile(current_vol: float, vol_history: list[float]) -> float:
    """Compute where current vol sits in history. Pure function."""
    if not vol_history:
        return 0.5

    below = sum(1 for v in vol_history if v < current_vol)
    return below / len(vol_history)

File: pypsik-live/src/risk/regime_detector.rs
/// Market regime detector — classifies into 4 regimes using statistical features.
///
/// Replaces numpy/scipy with pure Rust math for:
/// - Standard deviation, skewness, kurtosis
/// - Volatility percentile ranking
/// - Regime classification logic

/// Market regime classification.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MarketRegime {
    CalmTrending,
    VolatileTrending,
    MeanReverting,
    Crash,
}

/// Immutable regime detection result.
#[derive(Debug, Clone)]
pub struct RegimeResult {
    pub regime: MarketRegime,
    pub confidence: f64,
    pub volatility_percentile: f64,
    pub skewness: f64,
    pub kurtosis_val: f64,
    pub energy: f64,
}

/// Detects market regime using statistical features.
pub struct RegimeDetector {
    lookback: usize,
    min_samples: usize,
    returns: Vec<f64>,
    vol_history: Vec<f64>,
}

impl RegimeDetector {
    pub fn new(lookback: usize) -> Self {
        Self {
            lookback,
            min_samples: 30,
            returns: Vec::new(),
            vol_history: Vec::new(),
        }
    }

    /// Record a new return observation.
    pub fn update(&mut self, return_value: f64) {
        self.returns.push(return_value);
        if self.returns.len() > self.lookback * 2 {
            self.returns = self.returns.split_off(self.returns.len() - self.lookback * 2);
        }
    }

    /// Classify the current market regime.
    pub fn detect(&mut self) -> RegimeResult {
        let returns = if self.returns.len() >= self.lookback {
            self.returns[self.returns.len() - self.lookback..].to_vec()
        } else {
            self.returns.clone()
        };

        if returns.len() < self.min_samples {
            return RegimeResult {
                regime: MarketRegime::MeanReverting,
                confidence: 0.0,
                volatility_percentile: 0.5,
                skewness: 0.0,
                kurtosis_val: 0.0,
                energy: 1.0,
            };
        }

        let features = compute_features(&returns);

        // Track vol percentile
        self.vol_history.push(features.vol);
        if self.vol_history.len() > self.lookback {
            self.vol_history = self.vol_history.split_off(self.vol_history.len() - self.lookback);
        }
        let vol_percentile = compute_vol_percentile(features.vol, &self.vol_history);

        let (regime, confidence) = classify_regime(&features, vol_percentile);
        let energy = compute_energy(confidence, vol_percentile);

        RegimeResult {
            regime,
            confidence,
            volatility_percentile: vol_percentile,
            skewness: features.skew,
            kurtosis_val: features.kurt,
            energy,
        }
    }
}

struct Features {
    vol: f64,
    skew: f64,
    kurt: f64,
    mean: f64,
}

fn compute_features(returns: &[f64]) -> Features {
    if returns.len() < 20 {
        return Features { vol: 0.0, skew: 0.0, kurt: 0.0, mean: 0.0 };
    }

    let n = returns.len() as f64;
    let mean = returns.iter().sum::<f64>() / n;
    let variance = returns.iter().map(|r| (r - mean).powi(2)).sum::<f64>() / n;
    let vol = variance.sqrt();

    // Skewness (bias-corrected)
    let m3 = returns.iter().map(|r| (r - mean).powi(3)).sum::<f64>() / n;
    let skew = if variance > 0.0 {
        m3 / (variance.powf(1.5)) * ((n * (n - 1.0)).sqrt() / (n - 2.0))
    } else {
        0.0
    };

    // Excess kurtosis (Fisher, bias-corrected)
    let m4 = returns.iter().map(|r| (r - mean).powi(4)).sum::<f64>() / n;
    let kurt = if variance > 0.0 {
        let k = m4 / variance.powi(2) - 3.0;
        // Simple bias correction
        k * (n - 1.0) * (n - 1.0) / ((n - 2.0) * (n - 3.0))
    } else {
        0.0
    };

    Features { vol, skew, kurt, mean }
}

fn classify_regime(features: &Features, vol_percentile: f64) -> (MarketRegime, f64) {
    // Crash detection
    let is_crash = features.skew < -1.0
        || (features.mean < -0.015 && features.vol > 0.02);

    if is_crash {
        let confidence = if features.skew < -1.0 {
            (features.skew.abs() / 3.0).min(1.0)
        } else {
            (features.mean.abs() / 0.05).min(1.0)
        };
        return (MarketRegime::Crash, confidence);
    }

    let has_direction = if features.vol > 0.0 {
        features.mean.abs() > 0.3 * features.vol
    } else {
        false
    };

    if has_direction {
        if vol_percentile > 0.75 {
            let confidence = (0.6 + 0.4 * vol_percentile).min(1.0);
            return (MarketRegime::VolatileTrending, confidence);
        }
        let confidence = (0.6 + 0.4 * (1.0 - vol_percentile)).min(1.0);
        return (MarketRegime::CalmTrending, confidence);
    }

    let confidence = (0.5 + 0.3 * (1.0 - vol_percentile)).min(1.0);
    (MarketRegime::MeanReverting, confidence)
}

fn compute_energy(confidence: f64, vol_percentile: f64) -> f64 {
    (1.0 - confidence) * 0.6 + vol_percentile * 0.4
}

fn compute_vol_percentile(current_vol: f64, vol_history: &[f64]) -> f64 {
    if vol_history.is_empty() {
        return 0.5;
    }
    let below = vol_history.iter().filter(|&&v| v < current_vol).count() as f64;
    below / vol_history.len() as f64
}
